app/javascript/packs/outdoor_map_builder.js

```diff
@@ -162,6 +162,115 @@ console.log('[MAP] Layers built');
+    }
+    const innerBuildingLayer = buildingLayer.getLayers()[0];
+    const buildingInfo = innerBuildingLayer.feature;
+    return buildingInfo.properties.name && buildingInfo.properties.name.toLowerCase().indexOf(query.toLowerCase()) > -1;
+  })
+}
+
+function buildingAtLocation(location) {
+  // different representation of the position for the pointInPolygon-method
+  const locationArray = [location.lng, location.lat]
+  for(const buildingLayer of allBuildings()) {
+    const innerBuildingLayer = buildingLayer.getLayers()[0];
+    const buildingInfo = innerBuildingLayer.feature
+    const polygonCoordinates = buildingInfo.geometry.coordinates[0];
+    if(pointInPolygon(locationArray, polygonCoordinates)) return buildingLayer
+    // The next line ensures that the buildings are highlighted and properly geocoded,
+    // when they are being routed to. This is because we route to the center of the building.
+    if(innerBuildingLayer.getCenter().toBounds(1).contains(location)) return buildingLayer
+  }
+  return undefined
+}
+
+const originalGeocoder = new L.Control.Geocoder.nominatim()
+
+const HPIGeocoder = {
+  getHPISuggestions(query){
+    return buildingsByQuery(query).map((buildingLayer) => {
+      const innerBuildingLayer = buildingLayer.getLayers()[0];
+      const buildingInfo = innerBuildingLayer.feature;
+      return {
+        name: buildingInfo.properties.name,
+        center: innerBuildingLayer.getCenter(),
+        bbox: innerBuildingLayer.getBounds()
+      }
+    });
+  },
+
+  buildingInfoAtLocation(location) {
+    if(!location) return undefined;
+    const buildingLayer = buildingAtLocation(location)
+    if(!buildingLayer) return undefined;
+    const innerBuildingLayer = buildingLayer.getLayers()[0];
+    const buildingInfo = innerBuildingLayer.feature;
+    return {
+      name: buildingInfo.properties.name,
+      center: innerBuildingLayer.getCenter(),
+      bbox: innerBuildingLayer.getBounds()
+    }
+  },
+
+  /**
+   * Performs a geocoding query and returns the results to the callback in the provided context
+   * @param query the query
+   * @param cb the callback function
+   * @param context the this context in the callback
+   */
+  geocode(query, cb, context){
+    var that = this;
+    // This construction is used to append further suggestions to the ones given by the originalGeocoder
+    var callbackProxy = function(originalSuggestions) {
+      var hpiSuggestions = that.getHPISuggestions(query);
+      cb.call(context, hpiSuggestions.concat(originalSuggestions));
+    }
+    originalGeocoder.geocode(query, callbackProxy, context);
+  },
+  
+  /**
+   * Performs a geocoding query suggestion (this happens while typing) and returns the results to the callback in the provided context
+   * @param query the query
+   * @param cb the callback function
+   * @param context the this context in the callback
+   */
+  suggest(query, cb, context){
+    cb.call(context, this.getHPISuggestions(query));
```

Comment:
Shouldn't we be consistent here and either also include nominatim entries (e.g. at the bottom of the list?) otherwise, we could also delete from geocode, since that will never be used if I'm not mistaken.

Reply:
Nominatim just implements the geocode function which is called once the user entered their query and not the suggest function which is called after every keystroke. This is probably to cut down on api calls. We wanted to keep their design decision but implement suggest with at least our suggestions.